File: certpost/renewal.py
# ----------------------------------------------------------------------------------------
#   renewal.py
#   ----------
#
#   Background certificate renewal thread. Checks daily and renews certificates
#   that are within 30 days of expiry.
#
#   (c) 2026 WaterJuice — Released under the Unlicense; see LICENSE.
#
#   Version History
#   ---------------
#   Mar 2026 - Created
# ----------------------------------------------------------------------------------------

# ----------------------------------------------------------------------------------------
#   Imports
# ----------------------------------------------------------------------------------------

import datetime
import sys
import threading
import traceback
from .acme import AcmeClient
from .storage import Storage

# ----------------------------------------------------------------------------------------
#   Constants
# ----------------------------------------------------------------------------------------

_CHECK_INTERVAL = 86400  # 24 hours
_RENEWAL_WINDOW = 30 * 86400  # 30 days before expiry

# ----------------------------------------------------------------------------------------
#   Renewal Thread
# ----------------------------------------------------------------------------------------


# ----------------------------------------------------------------------------------------
class RenewalThread:
    """Background thread that checks and renews certificates."""

    # ------------------------------------------------------------------------------------
    #   Construction
    # ------------------------------------------------------------------------------------

    def __init__(self, storage: Storage, acme: AcmeClient) -> None:
        """Initialise the renewal thread."""
        self._storage = storage
        self._acme = acme
        self._stop_event = threading.Event()
        self._thread: threading.Thread | None = None

    # ------------------------------------------------------------------------------------
    #   Public methods
    # ------------------------------------------------------------------------------------

    # ------------------------------------------------------------------------------------
    def start(self) -> None:
        """Start the renewal thread."""
        self._thread = threading.Thread(
            target=self._run, daemon=True, name="cert-renewal"
        )
        self._thread.start()

    # ------------------------------------------------------------------------------------
    def stop(self) -> None:
        """Signal the renewal thread to stop."""
        self._stop_event.set()

    # ------------------------------------------------------------------------------------
    #   Private methods
    # ------------------------------------------------------------------------------------

    # ------------------------------------------------------------------------------------
    def _run(self) -> None:
        """Main loop — check for renewals, then sleep until next check."""
        # Initial delay to let the server start up
        if self._stop_event.wait(10):
            return

        while not self._stop_event.is_set():
            try:
                self._check_renewals()
            except Exception:
                print(
                    f"  [renewal] Error during renewal check:\n{traceback.format_exc()}",
                    file=sys.stderr,
                )

            # Sleep until next check (or until stopped)
            if self._stop_event.wait(_CHECK_INTERVAL):
                return

    # ------------------------------------------------------------------------------------
    def _check_renewals(self) -> None:
        """Check all domains and renew certificates that are near expiry."""
        domains = self._storage.get_domains()
        now = datetime.datetime.now(datetime.UTC)

        for domain in domains:
            subdomain = domain.get("subdomain", "")
            status = domain.get("status", "")
            expires_at_str = domain.get("cert_expires_at")

            if not subdomain:
                continue

            # Issue certs for pending domains
            if status == "pending":
                self._issue_cert(subdomain)
                continue

            # Renew certs within the renewal window
            if status == "issued" and expires_at_str is not None:
                expires_at = datetime.datetime.fromisoformat(str(expires_at_str))
                time_remaining = (expires_at - now).total_seconds()
                if time_remaining < _RENEWAL_WINDOW:
                    days_left = time_remaining / 86400
                    print(
                        f"  [renewal] Certificate for {subdomain} expires in {days_left:.0f} days, renewing...",
                        file=sys.stderr,
                    )
                    self._issue_cert(subdomain)

    # ------------------------------------------------------------------------------------
    def _issue_cert(self, subdomain: str) -> None:
        """Issue or renew a certificate for a subdomain."""
        try:
            self._storage.update_domain(subdomain, {"status": "issuing"})
            self._acme.issue_certificate(subdomain)
        except Exception as e:
            error_msg = f"{type(e).__name__}: {e}"
            print(
                f"  [renewal] Failed to issue cert for {subdomain}: {error_msg}",
                file=sys.stderr,
            )
            self._storage.update_domain(
                subdomain,
                {
                    "status": "error",
                    "last_error": error_msg,
                },
            )

File: internal/renewal/renewal.go
// ---------------------------------------------------------------------------------------
//
//	renewal.go
//	----------
//
//	Background certificate renewal goroutine. Checks daily, proactively renews
//	the 2 oldest certs per cycle to keep them fresh, with a 30-day expiry
//	safety net for forced renewal.
//
//	(c) 2026 WaterJuice — Released under the Unlicense; see LICENSE.
//
//	Version History
//	---------------
//	Apr 2026 - Created (Go rewrite)
//
// ---------------------------------------------------------------------------------------
package renewal

import (
	"context"
	"fmt"
	"sort"
	"time"

	"github.com/WaterJuice/certpost/internal/acme"
	"github.com/WaterJuice/certpost/internal/logbuf"
	"github.com/WaterJuice/certpost/internal/storage"
)

const (
	checkInterval     = 24 * time.Hour
	renewalWindow     = 30 * 24 * time.Hour // 30 days before expiry (forced renewal)
	dailyRenewalLimit = 2                   // max certs to proactively renew per cycle
)

// Start starts the renewal goroutine. Cancel the context to stop it.
func Start(ctx context.Context, s *storage.Storage, acmeClient *acme.Client) {
	go run(ctx, s, acmeClient)
}

func run(ctx context.Context, s *storage.Storage, acmeClient *acme.Client) {
	// Initial delay to let the server start up
	select {
	case <-time.After(10 * time.Second):
	case <-ctx.Done():
		return
	}

	for {
		checkRenewals(s, acmeClient)

		select {
		case <-time.After(checkInterval):
		case <-ctx.Done():
			return
		}
	}
}

type renewalCandidate struct {
	timeRemaining float64
	subdomain     string
}

func checkRenewals(s *storage.Storage, acmeClient *acme.Client) {
	defer func() {
		if r := recover(); r != nil {
			logbuf.Log("renewal", fmt.Sprintf("Error during renewal check: %v", r))
		}
	}()

	domains, err := s.GetDomains()
	if err != nil {
		logbuf.Log("renewal", fmt.Sprintf("Error reading domains: %v", err))
		return
	}

	now := time.Now().UTC()
	var proactiveCandidates []renewalCandidate

	for _, domain := range domains {
		subdomain, _ := domain["subdomain"].(string)
		status, _ := domain["status"].(string)
		expiresAtStr, _ := domain["cert_expires_at"].(string)

		if subdomain == "" {
			continue
		}

		// Issue certs for pending domains
		if status == "pending" {
			issueCert(s, acmeClient, subdomain)
			continue
		}

		// Force-renew certs within the expiry window (safety net)
		if status == "issued" && expiresAtStr != "" {
			expiresAt, err := time.Parse(time.RFC3339, expiresAtStr)
			if err != nil {
				continue
			}
			timeRemaining := expiresAt.Sub(now)
			if timeRemaining < renewalWindow {
				daysLeft := timeRemaining.Hours() / 24
				logbuf.Log("renewal", fmt.Sprintf(
					"Certificate for %s expires in %.0f days, renewing...", subdomain, daysLeft))
				issueCert(s, acmeClient, subdomain)
			} else {
				proactiveCandidates = append(proactiveCandidates, renewalCandidate{
					timeRemaining: timeRemaining.Seconds(),
					subdomain:     subdomain,
				})
			}
		}
	}

	// Proactively renew the oldest certs (earliest expiry first)
	sort.Slice(proactiveCandidates, func(i, j int) bool {
		return proactiveCandidates[i].timeRemaining < proactiveCandidates[j].timeRemaining
	})

	renewed := 0
	for _, c := range proactiveCandidates {
		if renewed >= dailyRenewalLimit {
			break
		}
		daysUntil := c.timeRemaining / 86400
		logbuf.Log("renewal", fmt.Sprintf(
			"Proactive renewal for %s (expires in %.0f days)", c.subdomain, daysUntil))
		issueCert(s, acmeClient, c.subdomain)
		renewed++
	}
}

func issueCert(s *storage.Storage, acmeClient *acme.Client, subdomain string) {
	_ = s.UpdateDomain(subdomain, map[string]any{"status": "issuing"})
	if err := acmeClient.IssueCertificate(subdomain); err != nil {
		errMsg := fmt.Sprintf("%v", err)
		logbuf.Log("renewal", fmt.Sprintf("Failed to issue cert for %s: %s", subdomain, errMsg))
		_ = s.UpdateDomain(subdomain, map[string]any{
			"status":     "error",
			"last_error": errMsg,
		})
	}
}
